docs-markdown/src/controllers/notebook-controller.ts

```diff
@@ -0,0 +1,207 @@
+	noActiveEditorMessage
+} from '../helper/common';
+import { resolve, join } from 'path';
+import { Selection, window, TextEditor } from 'vscode';
+import { spawn, execSync } from 'child_process';
+import { unlinkSync, writeFileSync } from 'fs';
+import { URL } from 'url';
+import { Command } from '../Command';
+
+export const notebookControllerCommands: Command[] = [
+	{ command: insertNotebook.name, callback: insertNotebook },
+	{ command: updateNotebook.name, callback: updateNotebook }
+];
+
+export async function insertNotebook() {
+	const editor = window.activeTextEditor;
+	if (!editor) {
+		noActiveEditorMessage();
+		return;
+	} else {
+		const url = await window.showInputBox({
+			prompt: 'Provide the GitHub URL of the *.ipynb notebook file'
+		});
+
+		if (!url) {
+			window.showInformationMessage('No URL to take action on.');
+			return;
+		}
+
+		const content = await getNotebookAsMarkdown(url);
+		if (content !== null) {
+			insertContentToEditor(editor, content, true);
+		}
+	}
+}
+
+export async function updateNotebook() {
+	const editor = window.activeTextEditor;
+	if (!editor) {
+		noActiveEditorMessage();
+		return;
+	} else {
+		const start = findLineNumberOfPattern(editor, '<!-- nbstart');
+		const end = findLineNumberOfPattern(editor, '<!-- nbend');
+
+		if (start > -1 && end > -1 && start < end) {
+			editor.selection = new Selection(start, 0, end, 14);
+			const url = getUrl(editor, start);
+			const content = await getNotebookAsMarkdown(url);
+			if (content) {
+				insertContentToEditor(editor, content, true);
+			}
+		} else {
+			window.showInformationMessage("There isn't a notebook to update in this document");
+		}
+	}
+}
+
+function getUrl(editor: TextEditor, start: number) {
+	const article = editor.document;
+	const text = article.lineAt(start).text;
+	const url = text.replace('<!-- nbstart ', '').replace(' -->', '');
+
+	return url;
+}
+
+function getConvertPromise(filePath: string) {
+	return new Promise<ConversionResult>((resolve, reject) => {
+		convertToMarkdown(filePath, (result: ConversionResult | null, errorOrNull: Error | null) => {
+			if (errorOrNull || result === null) {
+				reject(errorOrNull);
+			} else {
+				resolve(result);
+			}
+		});
+	});
+}
+
+async function getNotebookAsMarkdown(url: string) {
+	const rawUrl = toRaw(url);
+	const ipynbJson = await getRawNotebookJson(rawUrl);
+	const filePath = 'temp.ipynb';
+	writeFileSync(filePath, ipynbJson);
+	const fullPath = resolve(filePath);
+	try {
+		const result = await getConvertPromise(fullPath);
+		if (result !== null) {
+			return wrapMarkdownInTemplate(result.markdown, rawUrl);
+		}
+
+		return null;
+	} finally {
+		unlinkSync(fullPath);
+	}
+}
+
+async function getRawNotebookJson(url: string): Promise<string> {
+	const response = await fetch(url);
+	const data = await response.data;
+
+	return JSON.stringify(data);
+}
+
+function wrapMarkdownInTemplate(content: string, url: string) {
+	// Advance all the headings by one, avoid multiple H1s
```

Comment:
I think we should remove this, and leave it to the user to decide how to handle it if the document has multiple h1's.